agents/sorenson_dlm/dlm_agent.py

```diff
@@ -255,6 +255,49 @@ def set_current(self, session, params=None):
 
         return True, 'Set current to {}'.format(params['current'])
 
+    @ocs_agent.param('auto_acquire', default=False, type=bool)
+    @ocs_agent.param('acq_params', type=dict, default=None)
+    @ocs_agent.param('force', default=False, type=bool)
+    def init_dlm(self, session, params=None):
+        """init_dlm(auto_acquire=False, acq_params=None, force=False)
+        **Task** - Perform first time setup of the Sorenson DLM communication.
+
+        Parameters:
+            auto_acquire (bool, optional): Default is False. Starts data
+                acquisition after initialization if True.
+            acq_params (dict, optional): Params to pass to acq process if
+                auto_acquire is True.
+            force (bool, optional): Force initialization, even if already
+                initialized. Defaults to False.
+        """
+        if params is None:
+            params = {}
+
+        if self.initialized and not params.get('force', False):
+            self.log.info("DLM already initialized Returning...")
+            return True, "Already initialized"
+
+        
+        with self._lock.acquire_timeout(job='init') as acquired1, \
+                self._acq_proc_lock.acquire_timeout(timeout=0., job='init') \
+                as acquired2:
+            if not acquired1:
+                self.log.warn(f"Could not start init because "
+                              f"{self._lock.job} is already running")
+                return False, "Could not acquire lock"
+            if not acquired2:
+                self.log.warn(f"Could not start init because "
+                              f"{self._acq_proc_lock.job} is already running")
+                return False, "Could not acquire lock"
+        session.set_status('running')
```

Comment:
A couple things here. First, this doesn't actually initialize the DLM object, that's still being done in `__init__()`. That should be moved here.

This looks like it's based off of the Lakeshore 372, which is perhaps not the best example, it's a bit old. Check out the example here: https://ocs.readthedocs.io/en/develop/developer/agent_references/timeoutlock.html#release-and-reacquire

That example doesn't have this strange double lock context.

Also, I'm noticing that you have two locks, `self.lock` and `self._lock`. Get rid of one of these and use a single lock everywhere. Otherwise you might "have" one of the locks, then check the other one, meaning you're potentially communicating simultaneously with the DLM, which is what the lock is trying to prevent in the first place.

Reply:
Several questions:

> First, this doesn't actually initialize the DLM object, that's still being done in __init__(). That should be moved here.

What should be moved here? The guts of the `__init__()` function? If this is the case, where do I supply arguments like `ip_address` and `f_sample`, in `init_dlm()`? (I.E. `init_dlm(self, session, params=None, ip_address, port, f_sample=2.5)`)

> This looks like it's based off of the Lakeshore 372, which is perhaps not the best example, it's a bit old. Check out the example here: https://ocs.readthedocs.io/en/develop/developer/agent_references/timeoutlock.html#release-and-reacquire

The double lock context being the `acquired1` and `acquired2` locks? Should i just be inserting the code block from the example that start at `with self.lock.acquire_timeout(timeout=0, job='acq') as acquired:`

Reply:
> What should be moved here? The guts of the `__init__()` function? If this is the case, where do I supply arguments like `ip_address` and `f_sample`, in `init_dlm()`? (I.E. `init_dlm(self, session, params=None, ip_address, port, f_sample=2.5)`)

The DLM objection initialization is this block of the `__init__()` function: https://github.com/simonsobs/socs/blob/2b56f4bdbe2421680f6bdfec529412e1a64635a4/agents/sorenson_dlm/dlm_agent.py#L106-L111

> The double lock context being the `acquired1` and `acquired2` locks? Should i just be inserting the code block from the example that start at `with self.lock.acquire_timeout(timeout=0, job='acq') as acquired:`

This task is a "short action", so you need this block:
```
        with self.acquire_timeout(timeout=3, job='short_action') as acquired:
            if not acquired:
                print(f"Lock could not be acquired because it is held by {self.lock.job}")
                return False

            # Your code that performs the initialization here
```

Reply:
So where do I pass the  `ip_address` and `port` arguments for the `init_dlm` function? Should that go in `init_params`?

Reply:
Keep them as arguments to the class, and store them as class attributes. Then use them within `init_dlm` via `self.ip_address`. The Lakeshore 372 Agent does this for example within its `init` task: https://github.com/simonsobs/socs/blob/53758a1f5866fcb03b0525adadd01537fb5eaf2c/agents/lakeshore372/LS372_agent.py#L195